modules/fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/fs2/Channel.scala

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright (c) 2020 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.snowplow.enrich.fs2
+
+import java.time.Instant
+import java.nio.charset.StandardCharsets.UTF_8
+
+import cats.effect.Concurrent
+import fs2.Pipe
+
+import com.snowplowanalytics.snowplow.enrich.common.outputs.EnrichedEvent
+import com.snowplowanalytics.snowplow.badrows.{BadRow, Failure, Payload => BadRowPayload}
+import com.snowplowanalytics.snowplow.enrich.common.utils.ConversionUtils
+
+/** An item that can be sunk into any one of the enrichment sinks: Bad, Good, or Pii */
+sealed trait Channel[+Bad, +Good, +Pii]
+
+object Channel {
+
+  /** An item destined for the Bad Sink */
+  case class Bad[B](badRow: B) extends Channel[B, Nothing, Nothing]
+
+  /** An item destined for the Good Sink */
+  case class Good[G](event: G) extends Channel[Nothing, G, Nothing]
+
+  /** An item destined for the Pii Sink */
+  case class Pii[P](event: P) extends Channel[Nothing, Nothing, P]
+
+  type Parsed = Channel[BadRow, EnrichedEvent, EnrichedEvent]
+  type Serialized = Channel[Array[Byte], Array[Byte], Array[Byte]]
+
+  /** Sends a bad/good/pii item into the appropriate sink */
+  def sink[F[_]: Concurrent, B, G, P](
+    bad: Pipe[F, B, Unit],
+    good: Pipe[F, G, Unit],
+    pii: Pipe[F, P, Unit]
+  ): Pipe[F, Channel[B, G, P], Unit] =
+    _.observe(_.collect { case Bad(br) => br }.through(bad))
+      .observe(_.collect { case Good(event) => event }.through(good))
+      .observe(_.collect { case Pii(event) => event }.through(pii))
```

Comment:
I'm not at ease with the fact that the type of each element inside the `Stream` of `Channel` can change, and that we need to call 3 times `observe()` and use pattern matching on class types.

I would personally be in favor of having more granular control over the `Stream`s and not have a generic `Channel` that can wrap any type of data.

That might just be a matter of taste, I'm not sure about performance benefits.

Reply:
Let me try to make you feel more at ease with this!

Channel (I will rename it to Output) is basically a 3-way `Either`.  While `Either` has just `Left` and `Right`, our `Output` has 3 possible types `Bad`, `Good`, `Pii` - but apart from that they are the same thing.

Now FS2 has a function called `observeEither` [which is implemented exactly the same as above](https://github.com/typelevel/fs2/blob/2af0571de34cc11b6b2bafeef0baf487f0760ad8/core/shared/src/main/scala/fs2/Stream.scala#L3672-L3681).  It does pattern matching on the elements, and it calls `observe` for each possible pattern match.  I think this is the only way to direct a stream into several different sinks depending on the result of processing.

> I would personally be in favor of having more granular control over the Streams

I felt the same!  This was why I experimented with typing the sinks as `A => F[Unit]` instead of typing them as `Pipe[F, A, Unit]`.  I found the former let me have more control over exactly when to sink events; whereas the latter forced me to use this multi-`observe` pattern and the 3-way either.

Unfortunately my experiment with `A => F[Unit]` failed because things got messy with the file sink.

Can you think of any other way of send a stream of 3 different outputs into 3 different sinks?

Reply:
Thank you for the link for `observeEither` implementation, your implementation should be the optimal one then.

> Can you think of any other way of send a stream of 3 different outputs into 3 different sinks?

I can't. I guess the alternative would be to not have this `Output` class and to use multiple `observeEither` after each step (enrich, serialization with possibly size violation), but your approach is probably easier to read.